README.md
<div align="center">

  <h1><code>je</code></h1>

  <h3>
    <strong>Jcr Exchange</strong>
  </h3>

  <p>
    <img src="https://img.shields.io/github/workflow/status/devzbysiu/je/ci?style=for-the-badge" alt="CI status badge" />
    <a href="https://crates.io/crates/je">
      <img src="https://img.shields.io/crates/v/je?style=for-the-badge" alt="Crates.io version" />
    </a>
    <a href="https://codecov.io/gh/devzbysiu/je">
      <img src="https://img.shields.io/codecov/c/github/devzbysiu/je?style=for-the-badge&token=f2339b3de9e44be0a902458a669c1160" alt="Code coverage"/>
    </a>
    <a href="https://crates.io/crates/je">
      <img src="https://img.shields.io/crates/l/je?style=for-the-badge" alt="License"/>
    </a>
  </p>

  <h3>
    <a href="#about">About</a>
    <span> | </span>
    <a href="#installation">Installation</a>
    <span> | </span>
    <a href="#configuration">Configuration</a>
    <span> | </span>
    <a href="#license">License</a>
    <span> | </span>
    <a href="#contribution">Contribution</a>
  </h3>

  <sub><h4>Built with 🦀</h4></sub>
</div>

# <p id="about">About</p>

Small utility for uploading/downloading content to/from running AEM instance.

**It's intended to be used as an external tool for IntelliJ IDEA to make content synchronization easier.**

```bash
je 0.3.0
Jcr Exchange - easy download and upload files to and from JCR

USAGE:
    je [FLAGS] [OPTIONS] <SUBCOMMAND>

FLAGS:
    -d, --debug      If enabled, deployed to AEM packages are left intact (are not deleted) to allow investigation
    -h, --help       Prints help information
    -V, --version    Prints version information
    -v, --verbose    Enables logs: -v - enables INFO log level -vv - enables DEBUG log level

OPTIONS:
    -p, --profile <profile>    Profile selection

SUBCOMMANDS:
    get           Downloads content to local file system
    get-bundle    Downloads bundle (pack of crx paths) defined in config file
    help          Prints this message or the help of the given subcommand(s)
    init          Initializes configuration file
    put           Uploads content to AEM instance
    reinit        Rewrites the configuration file with newest version
```

# <p id="installation">Installation</p>

### Standard installation
- go to [releases](https://github.com/devzbysiu/je/releases) page
- download the latest `je` archive for your OS
- extract it
- include path to `je` executable into your PATH variable

### Rust programmers

#### Install
```bash
cargo install je
```

#### Upgrade
```bash
cargo install --force je
```

# <p id="configuration">Configuration</p>

### Default
Configuration file is **not** required. Without it, `je` will use default configuration.
However, you can still initialize config and change it. The default configuration is also the initial
one:

```
❯ je init
❯ cat .je
ignore_properties = []

[[profile]]
name = "author"
addr = "http://localhost:4502"
user = "admin"
pass = "admin"
```
### Customize
Here is more complex configuration with description of its fields:
```toml
ignore_properties = [{ type = "contains", value = "jcr:createdBy" },
                     { type = "regex", value = '.*=\[]' }]

[[profile]]
name = "author"
addr = "http://localhost:4502"
user = "admin"
pass = "admin"

[[profile]]
name = "publish"
addr = "http://localhost:4503"
user = "admin"
pass = "admin"

[[bundle]]
name = "configs"
files = ["/apps/my-app/config", "/config/my-app"]

[[bundle]]
name = "dam"
paths = ["/content/dam/my-app/thumbnails", "/content/dam/my-app/files"]
```

- `ignore_properties` - tell `je` which properties of `.content.xml` should be removed after
downloading the content; currently, two types of ignoring mechanisms are available:
  - `contains` - executes `line.contains(value)` on each line
  - `regex` - executes `regex.is_match(line)` on each line, it uses Perl-style regular expressions
- profile section - you can add multiple profiles, each with settings:
  - `name` - name of the profile, later it can be used with `--profile` option to specify which
    instance is the target; if not specified, the first profile from the config is used
  - `addr` - address of the instance, including port if domain is not available
  - `user` - user used to authenticate to AEM instance
  - `pass` - password used to authenticate to AEM instance
- bundles section - you can define packs of files which will be synchronized in one run:
  - `name` - name of the bundle, it can be later used with `--bundle` option to specify which
    file pack to synchronize
  - `paths` - which file paths are part of the bundle

> :warning: If you used older version of `je`, then you can transform the configuration file to the
new form. Details below.
<details>
  <summary>Details</summary>
  `je reinit` will change:
  
  ```toml
  ignore_properties = ["jcr:created", "jcr:createdBy"]

  # (...)
  ```
  to
  ```toml
  version = "0.3.0"

  [[ignore_properties]]
  type = "contains"
  value = "jcr:created"

  [[ignore_properties]]
  type = "contains"
  value = "jcr:createdBy"

  # (...)
  ```
  which is equivalent to:
  ```toml
  version = "0.3.0"

  ignore_properties = [{type = "contains", value = "jcr:created"},
                       {type = "contains", value = "jcr:createdBy"}]

  # (...)
  ```

  You can use any of those two formats. The difference is the result of <a href=https://github.com/alexcrichton/toml-rs/issues/265>this issue</a>.

</details>

### IntelliJ Setup

#### Add `je` commands:

1. Go to `Settings -> Tools -> External Tools`.
2. Add new external tool using `+` sign.
3. Configure the tool like on the screenshot below.

![je get configuration](./res/je-get.png)

Similarly add and configure `je put` command:

![je put configuration](./res/je-put.png)


##### Notes
- if you don't have `je` in PATH, you can set full path in `Program` input
- `Arguments` input:
  - `-vv` - sets verbose level, `-vv` means DEBUG log level, `-v` means INFO log level; it's optional
  - `--profile` - sets target profile; if not provided, first profile from config will be used
  - `-d` - sets debug mode in which temporary packages uploaded to AEM won't be deleted
    to allow validation during debugging; it's optional
  - subcommand (`put` or `get`)
  - `$FilePath$` - IntelliJ variable which will be substituted during command execution, its absolute
    path to a file on which command is executed
- `Working directory` input:
  - set to `$ProjectFileDir$` allows running `je` from the project root
  - keep in mind that `je` will try to read config file from the set working directory

# <p id="license">License</p>

This project is licensed under either of

- Apache License, Version 2.0, (LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0)
- MIT license (LICENSE-MIT or http://opensource.org/licenses/MIT)

at your option.

# <p id="contribution">Contribution</p>


Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.

See [CONTRIBUTING.md](./CONTRIBUTING.md) for hints on how to contribute to `je`.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.0] - 2021-29-07
### Added
- Mechanism for adjusting the old configuration file to the new version via `je reinit`
- Support for bundles (packs of files)
- Regex type mechanism for ignoring properties

### Changed
- [Fix](https://github.com/devzbysiu/je/commit/a405d0240562a1766c3aa60c7e541decfbb66af7) issue with cross-partition file move on windows: [#3](https://github.com/devzbysiu/je/issues/3)
- [Fix](https://github.com/devzbysiu/je/commit/ccb132618ea5047bfbbdd7c4eb26972bfe9aad64) issue with incorrect paths in zip file

## [0.2.0] - 2020-25-10
### Added

- Support for profiles via `--profile`.


src/cfgmgr.rs
use crate::cfg::{Cfg, IgnoreProp, IgnoreType, Instance};
use anyhow::Result;
use log::{debug, info};
use serde_derive::{Deserialize, Serialize};
use std::env;
use std::fs::read_to_string;
use std::path::Path;

pub(crate) const CONFIG_FILE: &str = ".je";

pub(crate) const CURRENT_VERSION: &str = env!("CARGO_PKG_VERSION");

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
pub(crate) struct Version {
    #[serde(rename = "version")]
    pub(crate) value: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
struct Pre030Cfg {
    pub(crate) ignore_properties: Vec<String>,

    #[serde(rename = "profile")]
    pub(crate) profiles: Vec<Instance>,
}

pub(crate) fn handle_cfg_load() -> Result<Cfg> {
    debug!("loading config: {:?}{}", env::current_dir(), CONFIG_FILE);
    if Path::new(CONFIG_FILE).exists() {
        let version: Version = toml::from_str(&read_to_string(CONFIG_FILE)?)?;
        if version.value.is_none() {
            let cfg: Pre030Cfg = toml::from_str(&read_to_string(CONFIG_FILE)?)?;
            Ok(convert_to_current_cfg(cfg))
        } else {
            // new configuration
            Ok(toml::from_str::<Cfg>(&read_to_string(CONFIG_FILE)?)?)
        }
    } else {
        debug!("{} config doesn't exists, loading default", CONFIG_FILE);
        Ok(Cfg::default())
    }
}

fn convert_to_current_cfg(cfg: Pre030Cfg) -> Cfg {
    info!("adjusting configuration to a newer version");
    let res = Cfg {
        version: Some(CURRENT_VERSION.to_string()),
        ignore_properties: adjust_ignore_props(cfg.ignore_properties),
        profiles: cfg.profiles,
        ..Cfg::default()
    };
    debug!("config after adjustments: {:#?}", res);
    res
}

fn adjust_ignore_props(props: Vec<String>) -> Vec<IgnoreProp> {
    let mut res = Vec::new();
    for prop in props {
        res.push(IgnoreProp {
            ignore_type: IgnoreType::Contains,
            value: prop,
        });
    }
    res
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::cfg::{Bundle, IgnoreProp, Instance};
    use crate::testutils::TestConfig;
    use anyhow::Result;

    #[test]
    fn test_handle_cfg_load_when_config_not_exists() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // given
        let expected_profiles = vec![Instance::new(
            "author",
            "http://localhost:4502",
            "admin",
            "admin",
        )];
        let expected_ignore_props = vec![];
        let expected_version = Some("0.3.0".into());

        // when
        let cfg = handle_cfg_load()?;

        // then
        assert_eq!(cfg.version, expected_version);
        assert_eq!(cfg.ignore_properties, expected_ignore_props);
        assert_eq!(cfg.profiles, expected_profiles);

        Ok(())
    }

    #[test]
    fn test_handle_cfg_load_when_config_exists_but_is_from_pre_030_version() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // given
        let test_config = TestConfig::new()?;
        let original_cfg_content = r#"ignore_properties = ["jcr:created", "jcr:createdBy"]

               [[profile]]
               name = "author"
               addr = "http://localhost:4502"
               user = "user1"
               pass = "pass1"
            "#;
        test_config.write_all(original_cfg_content)?;

        let expected_profiles = vec![Instance::new(
            "author",
            "http://localhost:4502",
            "user1",
            "pass1",
        )];
        let expected_ignore_props = vec![
            IgnoreProp {
                ignore_type: IgnoreType::Contains,
                value: "jcr:created".to_string(),
            },
            IgnoreProp {
                ignore_type: IgnoreType::Contains,
                value: "jcr:createdBy".to_string(),
            },
        ];
        let expected_version = Some("0.3.0".into());

        // when
        let cfg = handle_cfg_load()?;

        // then
        assert_eq!(cfg.version, expected_version);
        assert_eq!(cfg.ignore_properties, expected_ignore_props);
        assert_eq!(cfg.profiles, expected_profiles);
        // do not rewrite config event when version doesn't match
        assert_eq!(original_cfg_content, test_config.read_all()?);

        Ok(())
    }

    #[test]
    fn test_handle_cfg_load_when_config_is_not_available() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // when
        let cfg = handle_cfg_load()?;

        let expected_profiles = vec![Instance::new(
            "author",
            "http://localhost:4502",
            "admin",
            "admin",
        )];

        // then
        assert_eq!(cfg.ignore_properties, Vec::<IgnoreProp>::new());
        assert_eq!(cfg.profiles, expected_profiles);

        Ok(())
    }

    #[test]
    fn test_instance_with_existing_profile() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // given
        let test_config = TestConfig::new()?;
        let original_cfg_content = r#"version = "0.3.0"
               ignore_properties = [{type = "contains", value = "prop1"},
                                    {type = "contains", value = "prop2"}]

               [[profile]]
               name = "author"
               addr = "http://localhost:4502"
               user = "user1"
               pass = "pass1"

            "#;
        test_config.write_all(original_cfg_content)?;
        let expected_instance = Instance::new("author", "http://localhost:4502", "user1", "pass1");

        // when
        let cfg = handle_cfg_load()?;
        let instance = cfg.instance(Some(&String::from("author")));

        // then
        assert_eq!(instance, expected_instance);
        // config contains current version so it shouldn't be rewritten
        assert_eq!(original_cfg_content, test_config.read_all()?);
        Ok(())
    }

    #[test]
    fn test_instance_with_not_existing_profile() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // given
        let test_config = TestConfig::new()?;
        let original_cfg_content = r#"version = "0.3.0"
               ignore_properties = [{type = "contains", value = "prop1"},
                                    {type = "contains", value = "prop2"}]

               [[profile]]
               name = "author"
               addr = "http://localhost:4502"
               user = "user1"
               pass = "pass1"
            "#;
        test_config.write_all(original_cfg_content)?;
        let default_instance = Instance::new("author", "http://localhost:4502", "admin", "admin");

        // when
        let cfg = handle_cfg_load()?;
        let instance = cfg.instance(Some(&String::from("not-existing")));

        // then
        assert_eq!(instance, default_instance);
        // config contains current version so it shouldn't be rewritten
        assert_eq!(original_cfg_content, test_config.read_all()?);
        Ok(())
    }

    #[test]
    fn test_instance_when_no_profile_was_selected() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // given
        let test_config = TestConfig::new()?;
        let original_cfg_content = r#"version = "0.3.0"
               ignore_properties = [{type = "contains", value = "prop1"},
                                    {type = "contains", value = "prop2"}]

               [[profile]]
               name = "publish"
               addr = "http://localhost:4503"
               user = "user2"
               pass = "pass2"

               [[profile]]
               name = "author"
               addr = "http://localhost:4502"
               user = "user1"
               pass = "pass1"

            "#;
        test_config.write_all(original_cfg_content)?;
        let first_instance = Instance::new("publish", "http://localhost:4503", "user2", "pass2");

        // when
        let cfg = handle_cfg_load()?;
        let instance = cfg.instance(None);

        // then
        assert_eq!(instance, first_instance);
        // config contains current version so it shouldn't be rewritten
        assert_eq!(original_cfg_content, test_config.read_all()?);
        Ok(())
    }

    #[test]
    fn test_bundles_when_bundle_defined() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // given
        let test_config = TestConfig::new()?;
        let original_cfg_content = r#"version = "0.3.0"
               ignore_properties = [{type = "contains", value = "prop1"},
                                    {type = "contains", value = "prop2"}]

               [[profile]]
               name = "author"
               addr = "http://localhost:4502"
               user = "user1"
               pass = "pass1"

               [[bundle]]
               name = "simple"
               paths = ["file1", "file2"]
            "#;
        test_config.write_all(original_cfg_content)?;
        let expected_bundle = Bundle::new("simple", vec!["file1", "file2"]);

        // when
        let cfg = handle_cfg_load();
        debug!("result: {:?}", cfg);
        let cfg = handle_cfg_load()?;
        let bundle = cfg.bundle(Some("simple"));

        // then
        assert_eq!(expected_bundle, bundle);
        // config contains current version so it shouldn't be rewritten
        assert_eq!(original_cfg_content, test_config.read_all()?);
        Ok(())
    }

    #[test]
    fn test_bundles_when_multiple_bundles_defined() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // given
        let test_config = TestConfig::new()?;
        let original_cfg_content = r#"version = "0.3.0"
               ignore_properties = [{type = "contains", value = "prop1"},
                                    {type = "contains", value = "prop2"}]

               [[profile]]
               name = "author"
               addr = "http://localhost:4502"
               user = "user1"
               pass = "pass1"

               [[bundle]]
               name = "simple"
               paths = ["file1", "file2"]

               [[bundle]]
               name = "other"
               paths = ["file3", "file4"]
            "#;
        test_config.write_all(original_cfg_content)?;
        let expected_simple_bundle = Bundle::new("simple", vec!["file1", "file2"]);
        let expected_other_bundle = Bundle::new("other", vec!["file3", "file4"]);

        // when
        let cfg = handle_cfg_load()?;
        let simple_bundle = cfg.bundle(Some("simple"));
        let other_bundle = cfg.bundle(Some("other"));

        // then
        assert_eq!(expected_simple_bundle, simple_bundle);
        assert_eq!(expected_other_bundle, other_bundle);
        // config contains current version so it shouldn't be rewritten
        assert_eq!(original_cfg_content, test_config.read_all()?);
        Ok(())
    }

    #[test]
    fn test_bundles_when_no_bundle() -> Result<()> {
        let _ = pretty_env_logger::try_init();
        // given
        let test_config = TestConfig::new()?;
        let original_cfg_content = r#"version = "0.3.0"
               ignore_properties = [{type = "contains", value = "prop1"},
                                    {type = "contains", value = "prop2"}]

               [[profile]]
               name = "author"
               addr = "http://localhost:4502"
               user = "user1"
               pass = "pass1"
            "#;
        test_config.write_all(original_cfg_content)?;
        let expected_bundle = Bundle::default();

        // when
        let cfg = handle_cfg_load()?;
        let bundle = cfg.bundle(Some("not-existing"));

        // then
        assert_eq!(expected_bundle, bundle);
        // config contains current version so it shouldn't be rewritten
        assert_eq!(original_cfg_content, test_config.read_all()?);
        Ok(())
    }

    #[test]
    #[should_panic]
    fn test_bundles_when_bundle_broken() {
        let _ = pretty_env_logger::try_init();
        // given
        let test_config = TestConfig::new().unwrap();
        test_config
            .write_all(
                r#"version = "0.3.0"
                   ignore_properties = [{type = "contains", value = "prop1"},
                                        {type = "contains", value = "prop2"}]

                   [[profile]]
                   name = "author"
                   addr = "http://localhost:4502"
                   user = "user1"
                   pass = "pass1"

                   [[bundle]]
                   paths = ["file3", "file4"]
                "#,
            )
            .unwrap();

        // when
        let _not_important = handle_cfg_load().unwrap(); // should panic
    }
}
